ompi/mca/coll/base/coll_base_alltoall.c

```diff
@@ -235,15 +235,90 @@ int ompi_coll_base_alltoall_intra_pairwise(const void *sbuf, size_t scount,
     return err;
 }
 
-
-int ompi_coll_base_alltoall_intra_bruck(const void *sbuf, size_t scount,
-                                         struct ompi_datatype_t *sdtype,
-                                         void* rbuf, size_t rcount,
-                                         struct ompi_datatype_t *rdtype,
-                                         struct ompi_communicator_t *comm,
-                                         mca_coll_base_module_t *module)
+/*
+ * 
+ * Function: ompi_coll_base_alltoall_intra_k_bruck using O(logk(N)) steps
+ * Accepts:  Same arguments as MPI_Alltoall
+ * Returns: MPI_SUCCESS or error code
+ * 
+ * Description: This method extends ompi_coll_base_alltoall_intra_bruck to handle any
+ * radix k(k >= 2).
+ * 
+ *   Example on 6 ranks with k = 4
+ *    #     0      1      2      3      4      5
+ *         [00]   [10]   [20]   [30]   [40]   [50]
+ *         [01]   [11]   [21]   [31]   [41]   [51]
+ *         [02]   [12]   [22]   [32]   [42]   [52]
+ *         [03]   [13]   [23]   [33]   [43]   [53]
+ *         [04]   [14]   [24]   [34]   [44]   [54]
+ *         [05]   [15]   [25]   [35]   [45]   [55]
+ *   After local rotation
```

Comment:
Why do you really need the local rotation ? I understand it makes the code easier to maintain but in has a significant and finally unnecessary cost, because at the end you are building the datatype by hand without taking advantage of it's continuity in memory. 

Second remark is related to the cost of creating and committing the datatype. I'm almost certain that this cost is expensive, especially for the middle range messages where the k-ary bruck is supposed to behave best. The result is that you pay a high cost to prepare the datatype, resulting in a non contiguous datatype while leads to lower performance communications (because non-contiguous data usually lead to copy in/out protocols). If instead of building the datatype you copy the data into a contiguous buffer, you avoid the cost of the datatype construction and communicate from contiguous to contiguous memory, with better outcome. The only potential drawback is the extra local copy before the send (similar to a pack).

Reply:
@jiaxiyan Please update the comment section after removing the rotation step(s).